src/clean_companies_data.py

```diff
@@ -0,0 +1,169 @@
+import logging
+import os
+import pandas as pd
+import numpy as np
+import matplotlib.pyplot as plt
+import missingno as msno
+import collections
+import itertools
+from tqdm import tqdm
+import nltk
+from datetime import datetime, timedelta
+import csv
+
+# Import and inspect initial data on companies.
+company_df = pd.read_csv('./data/external/companies.csv')
+
+# Visualize missing values in data.
+msno.matrix(company_df)
+
+# Drop if market is empty
+company_df = company_df.dropna(subset = ['market'])
+
+# Basic dataframe shape and columns.
+company_df.shape
+
+company_df.columns
+
+# Keep columns that are relevant.
+company_df = company_df[['permalink', 'name', 'market', 'category_list', 'funding_total_usd', 'status', 'country_code', 'funding_rounds', 'founded_at', 'founded_month', 'founded_quarter', 'founded_year', 'first_funding_at', 'last_funding_at']]
+
+# Lower case if a column is of string type.
+company_df = company_df.apply(lambda column: column.str.lower() if pd.api.types.is_string_dtype(column) else column)
+
+# Check number of unique market types.
+len(company_df['market'].unique())
+
+# Save list of all market types.
+market_types = list(company_df['market'].unique())
+
+# Save list of top 50 market types by company counts.
+top_50_markets = list(company_df['market'].value_counts().nlargest(50).index)
+
+outside_50_markets = [x for x in market_types if x not in top_50_markets]
+
+# Create dictionary of top 50 markets.
+top_50_dict = dict.fromkeys(top_50_markets)
+
+# First pass at collapsing market categories.
+# Done by checking if first word in a category is present in other category labels.
+# Example 1: word "advertising" apppears in "advertising platforms".
+# Example 2: "enterprise" is the first word in "enterprise software" apppears in "enterprise 2.0"
+
+for key in tqdm(top_50_dict.keys()):
+    # Initialize each key with empty list
+    top_50_dict[key] = []
+    for cat in outside_50_markets:
+        # Check if first word is present in category, add to key value list, remove from category list.
+        if key.split(' ')[0] in cat.split(' '):
+            top_50_dict[key].append(cat)
+            outside_50_markets.remove(cat)
+        elif (len(key.split(' ')) > 1) and (key.split(' ')[1] == 'and') and (key.split(' ')[2] in cat.split(' ')):
+            top_50_dict[key].append(cat)
+            outside_50_markets.remove(cat)
+
+# Check how many categories are left.
+remainder_outside_50 = outside_50_markets
+len(remainder_outside_50)
+
+# Second pass at collapsing market categories.
+# Market category labels are split into character n-grams.
+# N-gram similarity Dice metric is calculated with intersection/union of n-gram sets.
```

Comment:
I'm not real familiar with this metric, but you might document why you use this/what it implies. Did you pick this over other options like Levensthein or jaccard distance for any particular reason?

By the looks of it, it seems expensive to compute? A  N-character word becomes a N-4+1 length list and then unioned with another list from different word? set operations are fairly expensive at larger sizes right?